rak811.go

```diff
@@ -162,40 +163,64 @@ func (l *Lora) SetConfig(config string) (string, error) {
 	return l.tx(fmt.Sprintf("set_config=%v", config))
 }
 
-// Get LoRaWAN configuration
+// GetConfig LoRaWAN configuration
 func (l *Lora) GetConfig(key string) (string, error) {
 	return l.tx(fmt.Sprintf("get_config=%s", key))
 }
 
-// Get LoRaWAN band region
+// GetBand LoRaWAN band region
 func (l *Lora) GetBand() (string, error) {
 	return l.tx("band")
 }
 
-// Set LoRaWAN band region
+// SetBand LoRaWAN band region
 func (l *Lora) SetBand(band string) (string, error) {
 	return l.tx(fmt.Sprintf("band=%s", band))
 }
 
-// JoinOTAA join the configured network in OTAA mode
-func (l *Lora) JoinOTAA() (string, error) {
+const (
+	// JoinSuccess successfull join.
+	JoinSuccess = "at+recv=3,0,0"
+	// JoinFail incorrect join config parameters.
+	JoinFail = "at+recv=4,0,0"
+	// JoinTimeout no response from a gateway.
+	JoinTimeout = "at+recv=6,0,0"
+)
+
+// JoinOTAA join the configured network in OTAA mode.
+// The module doesn't accept any other command before it returns a response.
+// Response: JoinSuccess, JoinFail, JoinTimeout
+// Returns an error when the timeout argument expires.
+// This is as a fail safe so that the caller can reset the module
+// if it doesn't return anything for a long period.
+func (l *Lora) JoinOTAA(timeout time.Duration) (string, error) {
 	if _, err := l.tx("join=otaa"); err != nil {
 		return "", err
 	}
-
-	// Wait for the registration success event.
-	resp, err := bufio.NewReader(l.port).ReadString('\n')
-	if err != nil {
-		if err == io.EOF {
-			return "", ErrTimeout
+	var (
+		resp  string
+		err   error
+		start = time.Now()
+	)
+	for {
```

Comment:
the reader should be created once outside the loop

Reply:
updated